Improving test coverage and small fixes for xplat locals

mishra14 opened the pull request:
PR contains code for the following - 
1. Improve test coverage for xplat locals by adding test cases for --clear | -c and adding Utils.cs to get the dotnet cli path, etc.
2. Add the troubleshooting link to the Package Manager Settings UI.
3. Better error detection in LocalsCommandRunner.
4. Enable Help option in xplat locals.

Fixes :  https://github.com/NuGet/Home/issues/3348 https://github.com/NuGet/Home/issues/3347

//cc : @joelverhagen @rohit21agrawal @emgarten @jainaashish @drewgil

Commit:
1. Adding help to xplat locals 2. converting locals xplat tests to command runner basec

Commit:
Adding method to fetch dotnet cli through util

Commit:
updating build.sh to build xplat dll before running tests

Commit:
mocking env variables for --clear option in xplat locals test cases

Commit:
changing test path to release

Commit:
changing dotnet build command

Commit:
Switching to release config to test

Commit:
1. Fixing troubleshoot link in UI to a FW link. 2. Adding errror check to localsCommandRunner.

Commit:
Adding documentation to Util.cs and removing stray changes

joelverhagen commented on src/NuGet.Core/NuGet.Commands/LocalsCommandRunner.cs:
```diff
@@ -31,9 +31,9 @@ private enum LocalResourceName
         private const string TempResourceName = "temp";
 
-        public bool Clear { get; set; }
+        private bool Clear { get; set; }
```
If private, why not a field?

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+
+namespace NuGet.XPlat.FuncTest
+{
+    internal class Util
```
Too generic, `DotnetCliUtil` perhaps

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+
+        private static string XplatDllShell { get; set; }
+
+        static Util()
```
No need for a static ctor or properties here, just set these strings to be `private const string`.

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+            DotnetCliBinary = @"dotnet";
+            DotnetCliExe = @"dotnet.exe";
+            XplatDll = @"NuGet.Core\NuGet.CommandLine.XPlat\bin\release\netcoreapp1.0\NuGet.CommandLine.XPlat.dll";
+            XplatDllShell = @"NuGet.Core/NuGet.CommandLine.XPlat/bin/release/netcoreapp1.0/NuGet.CommandLine.XPlat.dll";
```
Same path just with different slashes? Use `static readonly string` with `Path.Combine` so you only need one field.

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+        /// </returns>
+        public static string GetDotnetCli()
+        {
+            var currentDirInfo = new DirectoryInfo(Directory.GetCurrentDirectory());
```
Can Xunit or something else tell you the location of the test assembly and use that as a starting point? `Directory.GetCurrentDirectory` is just another static that can be modified by another test.
See:
https://docs.microsoft.com/en-us/dotnet/core/api/system.reflection.assembly#System_Reflection_Assembly_Location

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+            while (parentDirInfo != null)
+            {
+                var dotnetCli = "";
+                foreach (var dir in parentDirInfo.EnumerateDirectories())
```
declare at inner scope

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+        /// Adds a few dummy text files at the specified path for testing nuget locals --clear
+        /// </summary>
+        /// <param name="path">Path which needs to be populated with dummy files</param>
+        public static void createTestFiles(string path)
```
Capitalize first letter.

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+                var xplatDll = "";
+                foreach (var dir in parentDirInfo.EnumerateDirectories())
+                {
+                    if (StringComparer.OrdinalIgnoreCase.Equals(dir.Name, "src"))
```
Why ignore case?

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+            var parentDirInfo = currentDirInfo.Parent;
+            while (parentDirInfo != null)
+            {
+                var xplatDll = "";
```
declare at inner scope

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+        {
+            Assert.True(
+                result.Item1 == 0,
+                "nuget.exe DID NOT SUCCEED: Ouput is " + result.Item2 + ". Error is " + result.Item3);
```
Might not be nuget.exe, right?

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+        {
+            Assert.True(
+                result.Item1 != 0,
+                "nuget.exe DID NOT FAIL: Ouput is " + result.Item2 + ". Error is " + result.Item3);
```
Same as above, might not be nuget.exe right?

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Util.cs:
```diff
@@ -0,0 +1,151 @@
+                result.Item1 == 0,
+                "nuget.exe DID NOT SUCCEED: Ouput is " + result.Item2 + ". Error is " + result.Item3);
+
+            if (!string.IsNullOrEmpty(expectedOutputMessage))
```
When would this be an empty string? And why treat that differently? Contains handles it.

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -26,19 +28,49 @@ public class XPlatLocalsTests
-        // [InlineData("locals --clear all")]
-        // [InlineData("locals -c all")]
-        public static void Locals_Succeeds(String args)
+        [InlineData("locals --clear all")]
```
:+1:

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -26,19 +28,49 @@ public class XPlatLocalsTests
+                {
+                    { "NUGET_PACKAGES", mockGlobalPackagesDirectory.FullName },
+                    { "NUGET_HTTP_CACHE_PATH", mockHttpCacheDirectory.FullName },
+                    { "TMP", mockTmpCacheDirectory.FullName }
```
`TMP` should work on Windows. It also works on Mac and Linux as expected?

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -47,53 +79,53 @@ public static void Locals_Succeeds(String args)
             // Assert
-            Assert.Equal("usage: NuGet locals <all | http-cache | global-packages | temp> [--clear | -c | --list | -l]" + Environment.NewLine + "For more information, visit http://docs.nuget.org/docs/reference/command-line-reference", log.ShowErrors());
-            Assert.Equal(1, exitCode);
+            Util.VerifyResultFailure(result, "error: usage: NuGet locals <all | http-cache | global-packages | temp> [--clear | -c | --list | -l]" + Environment.NewLine + "error: For more information, visit http://docs.nuget.org/docs/reference/command-line-reference");
```
string concat to split this into shorter lines

joelverhagen commented:
:watch:

Commit:
Fixing code based on PR review

mishra14 commented:
@joelverhagen : Addressed the comments. 
Major ones - 
1. Made all fields in LocalsCommandRunner.cs private, since they are not used outside.
2. Renamed Utils and changed strings to const.
3. Unified dll path using Path.combine.
4. Changed env variable for temp directory.

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -26,19 +29,50 @@ public class XPlatLocalsTests
+                    {
+                        { "NUGET_PACKAGES", mockGlobalPackagesDirectory.FullName },
+                        { "NUGET_HTTP_CACHE_PATH", mockHttpCacheDirectory.FullName },
+                        { RuntimeEnvironmentHelper.IsWindows?"TMP":"TMPDIR", mockTmpCacheDirectory.FullName }
```
nit: space

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -26,19 +29,50 @@ public class XPlatLocalsTests
+                // Unix uses TMPDIR as environment variable as opposed to TMP on windows
+
+                // Assert
+                DotnetCliUtil.VerifyResultSuccess(result, string.Empty);
```
You should verify the files are gone, right?

joelverhagen replied:
`!Directory.Exists` and `Directory.EnumerateFileSystemEntries` recursive count should be 0.

joelverhagen commented:
Your tests need to assert the user scenario. `--list` is just printing stuff out (right? no IO?) so that's easier. But you should have assertions verifying that clear of each kind of directory actually empties (or deletes) the directory in question. :watch:

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -47,53 +81,66 @@ public static void Locals_Succeeds(String args)
         {
             // Arrange
-            var log = new TestCommandOutputLogger();
+            var expectedResult = string.Concat("error: usage: NuGet locals <all | http-cache | global-packages | temp> [--clear | -c | --list | -l]",
```
why is usage coming out as an error log?

joelverhagen replied:
Spoke offline about this one. Seems like this is consistent with the other commands.

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -47,53 +81,66 @@ public static void Locals_Succeeds(String args)
-            var log = new TestCommandOutputLogger();
+            var expectedResult = string.Concat("error: usage: NuGet locals <all | http-cache | global-packages | temp> [--clear | -c | --list | -l]",
+                                               Environment.NewLine,
+                                               "error: For more information, visit http://docs.nuget.org/docs/reference/command-line-reference");
```
you should have specific messages about what is missing from the `args` and assertions on those

joelverhagen commented:
What's the plan for introducing these tests to nuget.exe? @drewgil is working on this for other commands.

Commit:
Adding end2end assertion in test cases

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -66,11 +100,19 @@ public static void Locals_Succeeds(string args)
+                if (StringComparer.Ordinal.Equals(cacheType, "all"))
+                {
+                    Assert.False(Directory.Exists(mockGlobalPackagesDirectory.FullName));
+                    var x = Directory.EnumerateFileSystemEntries(mockGlobalPackagesDirectory.FullName);
```
What's this `x`?

mishra14 replied:
ah shoot. test code! sorry about that. :( Fixing it now!

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -66,11 +100,19 @@ public static void Locals_Succeeds(string args)
                 // Unix uses TMPDIR as environment variable as opposed to TMP on windows
 
                 // Assert
+                if (StringComparer.Ordinal.Equals(cacheType, "all"))
```
Conditions in a test should be thought through carefully (since multiple code paths are typically what need the testing!). You should test each locals directory individually wrt to the `cacheType` you parse out. For example, if `--clear global-packages` you would expected the `temp` local to _not_ be cleared, right?

nit: you can just `cacheType == "all"` here for brevity.

Commit:
Fixing wrong file in previous commit

Commit:
Adding more condition checks

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -100,19 +101,29 @@ public static void Locals_Clear_Succeeds(string args)
                 }
+                else if (StringComparer.Ordinal.Equals(cacheType, "global-packages"))
+                {
+                    Assert.False(Directory.Exists(mockGlobalPackagesDirectory.FullName));
```
Assert that the test files created above `DotnetCliUtil.CreateTestFiles` were not deleted. Basically, in this particular case, the `temp` and `http-cache` directories should have the same file system entires as before running the command.

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -82,7 +82,8 @@ public static void Locals_Clear_Succeeds(string args)
                 var mockHttpCacheDirectory = Directory.CreateDirectory(Path.Combine(mockBaseDirectory.Path, @"http-cache"));
-                var mockTmpCacheDirectory = Directory.CreateDirectory(Path.Combine(mockBaseDirectory.Path, @"temp"));
+                var mockTmpDirectory = Directory.CreateDirectory(Path.Combine(mockBaseDirectory.Path, @"temp"));
+                var mockTmpCacheDirectory = Directory.CreateDirectory(Path.Combine(mockBaseDirectory.Path, @"NuGetScratch"));
```
I see 4 directories here. 3 of which get populated, what's the difference between `mockTmpDirectory` and `mockTmpCacheDirectory`?

mishra14 replied:
mockTmpDirectory is a mock for c:/Users/..../Temp
mockTmpCacheDirectory is a mock for c:/Users/..../Temp/NuGetScratch

mockTmpCacheDirectory is contained within mockTmpDirectory, to match the folder structure as in actual Temp directory.

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -26,19 +29,113 @@ public class XPlatLocalsTests
-            Assert.Equal(string.Empty, log.ShowErrors());
-            Assert.Equal(0, exitCode);
+                // Act
+                var result = CommandRunner.Run(
```
What happens if the running process does not have access to delete a file or directory in one of these locals directories? Does it fail in the middle of the recursive delete? Does it skip that inaccessible file?

Commit:
Fixing path

Commit:
Fixing error messages and test cases

Commit:
Adding validation to ensure cleared directories were deleted

Commit:
Fixing an error message in test cases

Commit:
Fixing error message, again!

joelverhagen commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -26,74 +29,242 @@ public class XPlatLocalsTests
+                // Assert
+                if (cacheType == "all")
+                {
+                    Assert.False(Directory.Exists(mockGlobalPackagesDirectory.FullName));
```
nit: you can use VerifyClearSuccess on these three, right?

joelverhagen commented:
rebase, respond to nit, :shipit:

Commit:
Removing stray space from error message

Commit:
Removing reflection used for fetching current directory

joelverhagen commented:
:shipit:

mishra14 commented:
Merged : https://github.com/NuGet/NuGet.Client/commit/89a4f21a14e632c60932339e834e8dd4aa30db5e
